Move to sqs

chaws opened the pull request:
This PR tweaks lots of files, so I tried to organize the changes by commits in the following order:

* 1st: (https://github.com/Linaro/qa-reports.linaro.org/commit/8fd42bb1e71db43a3017690eab754bb272ad0199) reproduce the existing `qa_reports_instance_profile` instance profile into terraform scripts, then change their name so that there'll be one per environment
* 2nd: (https://github.com/Linaro/qa-reports.linaro.org/commit/7b59c15e6089b94d51b9de335f548ef224e5d0f1) remove terraform setup for rabbitmq and make ansible set it up only for development environments
* 3rd: (https://github.com/Linaro/qa-reports.linaro.org/commit/2f2cf904c022afae557e63754d1db242bf27b220) do the actual migration to AWS SQS
  * pre-create queues prefixed with environment name
  * attach SQS queue access to the environment role so that workers and webservers can talk to SQS without explicit credentials
  * use `list_sqs_queues.py` instead of `rabbitmqctl list_queues` so that munin can monitor number of tasks in queues

@danrue , could you help us make sure things are reasonable in terraform? I managed to understand `instance_profile` OK, but it turned out that we already had an existing one named `qa_reports_instance_profile`, I don't know who created that. I think it got it right in reproducing its permissions into terraform.

@terceiro , is it OK for `list_sqs_queues.py` to live under `/etc/munin/plugins`? I had to write our own script, since the solution we're using to authenticate to SQS would break that plugin I told you in the meeting. 

I wanted to run this in staging after we're sure of the things above.

chaws committed:
terraform: add `worker_type` when saving ansible inventory

chaws committed:
terraform: manage qa_reports instance role in terraform

Makes sure that roles needed by ec2 instances are created
and managed within terraform.

chaws committed:
ansible/terraform: setup rabbitmq only for dev environment

chaws committed:
ansible/terraform: use sqs for staging and production

danrue commented on ansible/munin/list_sqs_queues.py:
```diff
@@ -0,0 +1,25 @@
+# ref: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sqs.html#SQS.Client.get_queue_attributes
+num_of_msgs_attr = 'ApproximateNumberOfMessages'
+
+environment = os.getenv('ENV', 'staging')
```
I guess this should be a host variable filled in from ansible?

chaws replied:
You're right, makes more sense to have it filled by ansible.

danrue commented on terraform/modules/sqs/sqs.tf:
```diff
@@ -0,0 +1,53 @@
+                "sqs:GetQueueUrl"
+            ],
+            "Resource": [
+                "arn:aws:sqs:${var.region}:${var.account_number}:${var.environment}_*"
```
ok I hate this. Why can't you use the list of queues that come from aws_sqs_queue.qa_reports_queue? Having to pass in region and account_number is smelly.

chaws replied:
I hate it as much as you do! I just couldn't figure it out. Here's what I attempted:

1st:
```
"Resource": ["${aws_sqs_queue.qa_reports_queue.*.arn}"]
```

2nd:
```
"Resource": "${aws_sqs_queue.qa_reports_queue.*.arn}"
```

3rd:
```
locals {
  arns = ["${aws_sqs_queue.qa_reports_queue.*.arn}"]
}
...
"Resource": "${local.arns}"
```

all attempts above resulted in the error below:
```
* module.sqs.aws_iam_role_policy.sqs_manage_policy: 1 error(s) occurred:
* module.sqs.aws_iam_role_policy.sqs_manage_policy: At column 1, line 1: output of an HIL expression must be a string, or a single list (argument 2 is TypeList)
```

After some googling, my last attempt was this stackoverflow answer: https://stackoverflow.com/a/45816492. It loads the policy straight out of a json file, instead of using here doc syntax. But that throws a syntax error and I didn't debug it to see what happened exactly.

I tried using terraform data source as well, but it fails because queues don't yet exit.

Any hints?

danrue replied:
I'm not sure, but perhaps join() can help? I think `aws_sqs_queue.qa_reports_queue` is a list, right?

```
"Resource": [ "${join(", ", aws_sqs_queue.qa_reports_queue)}" ]
```

chaws replied:
Awesome! I hadn't thought of `join`. After minor tweaking, it ended up working with `["${join("\",\"", aws_sqs_queue.qa_reports_queue.*.arn)}"]`

danrue reviewed (approved):
Looks good for the most part - the instance profile part seems right to me too. I left a couple comments.

terceiro commented on terraform/scripts/state_to_ansible_inventory.py:
```diff
@@ -47,12 +47,17 @@
+        worker_quick = ' worker_type=quick'
+    else:
+        worker_quick = ''
+    print('{} ansible_host={}{}'.format(host['name'], host['ip'], worker_quick))
```
this is kind of a hack. now that we seem to have solved the performance issues, maybe there is no need for this separation among workers anymore?

chaws replied:
hum, I still think it's a good thing to have a separate queue for fetch tasks due to some test jobs that still takes lots of time to process, e.g. android-lkft. I was thinking about going back to two workers only: one for fetching and another one for anything else...

terceiro reviewed (commented):
LGTM overall. I have a small comment though

chaws committed:
terraform: fix certificate deletion

Prevents terraform from deleting certificates that were created with
`validation=NONE` and later set to `validation=EMAIL`.

chaws committed:
terraform: rds: enables autoscaling up to 500G

chaws committed:
ansible: munin: squad_job_count always exit with 0

Munin log was complaining that `squad_job_count` exited
with status `1`.

chaws commented:
Just for the record: I've updated the infrastructure in staging to use SQS and it's working alright. I've submitted a few jobs and everything seems to be running normally. I'd say we leave it running over the weekend until mid next week then replicate these changes in production.

BTW: here's a list of things that are going to change in production. All EC2 instances will be destroyed and recreated:

```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

aws_lb_target_group.qa-reports-tg: Refreshing state... (ID: arn:aws:elasticloadbalancing:us-east-1:...etgroup/qa-reports-tg/0825c86d0387f9fd)
aws_acm_certificate.acm-cert: Refreshing state... (ID: arn:aws:acm:us-east-1:495570029050:cert...e/23ee8d50-7224-49e2-b535-faea6b239b94)
aws_sns_topic.qa_reports_cloudwatch_notifications: Refreshing state... (ID: arn:aws:sns:us-east-1:495570029050:prod...on_qa_reports_cloudwatch_notifications)
aws_db_subnet_group.default: Refreshing state... (ID: productionqareports)
aws_db_parameter_group.default: Refreshing state... (ID: production-qa-reports-postgresql-params)
aws_key_pair.auth: Refreshing state... (ID: qa-reports-production)
aws_security_group.qa-reports-lb-sg: Refreshing state... (ID: sg-0051807f9c7df9db8)
data.aws_subnet.oursubnets[1]: Refreshing state...
data.aws_subnet.oursubnets[0]: Refreshing state...
aws_security_group.qa-reports-ec2-worker: Refreshing state... (ID: sg-0b4879d45d89a3d0b)
aws_security_group.qa-reports-ec2-www: Refreshing state... (ID: sg-0784d01ac183ae01c)
aws_lb.qa-reports-lb: Refreshing state... (ID: arn:aws:elasticloadbalancing:us-east-1:...cer/app/qa-reports-lb/992ac4ad50fa3c61)
aws_instance.qa-reports-www[0]: Refreshing state... (ID: i-07afe5562c94db883)
aws_instance.qa-reports-www[1]: Refreshing state... (ID: i-0c703fe5a346fb02f)
aws_instance.qa-reports-worker[2]: Refreshing state... (ID: i-07ea0b70b54c08ace)
aws_instance.qa-reports-worker[1]: Refreshing state... (ID: i-07df711da0c6e9320)
aws_instance.qa-reports-worker[0]: Refreshing state... (ID: i-0ebd191d546eac7af)
aws_security_group.qa-reports-db-sg: Refreshing state... (ID: sg-041f059c94273e34f)
aws_lb_listener.qa-reports-lb-listener-80: Refreshing state... (ID: arn:aws:elasticloadbalancing:us-east-1:...s-lb/992ac4ad50fa3c61/e50049ddd30873cd)
aws_route53_record.qa-reports-lb-dns: Refreshing state... (ID: Z27NRA2FV79C84_qa-reports.ctt.linaro.org_A)
aws_lb_listener.qa-reports-lb-listener-443: Refreshing state... (ID: arn:aws:elasticloadbalancing:us-east-1:...s-lb/992ac4ad50fa3c61/a58ea6e93c8c44e0)
aws_db_instance.default: Refreshing state... (ID: terraform-20180420145948581500000001)
aws_lb_target_group_attachment.qa-reports-www-lb[1]: Refreshing state... (ID: arn:aws:elasticloadbalancing:us-east-1:...86d0387f9fd-20180420150016829100000003)
aws_lb_target_group_attachment.qa-reports-www-lb[0]: Refreshing state... (ID: arn:aws:elasticloadbalancing:us-east-1:...86d0387f9fd-20180420150016760500000002)
aws_cloudwatch_metric_alarm.www_disk_usage[1]: Refreshing state... (ID: qa-reports-www-1_disk_capacity)
aws_cloudwatch_metric_alarm.www_disk_usage[0]: Refreshing state... (ID: qa-reports-www-0_disk_capacity)
aws_cloudwatch_metric_alarm.worker_disk_usage[1]: Refreshing state... (ID: qa-reports-worker-1_disk_capacity)
aws_cloudwatch_metric_alarm.worker_disk_usage[0]: Refreshing state... (ID: qa-reports-worker-0_disk_capacity)
aws_cloudwatch_metric_alarm.worker_disk_usage[2]: Refreshing state... (ID: qa-reports-worker-2_disk_capacity)

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
  ~ update in-place
-/+ destroy and then create replacement

Terraform will perform the following actions:

  + aws_iam_instance_profile.qa_reports_instance_profile
      id:                                    <computed>
      arn:                                   <computed>
      create_date:                           <computed>
      name:                                  "production_qa_reports_instance_profile"
      path:                                  "/"
      role:                                  "production_qa_reports_instance_iam_role"
      roles.#:                               <computed>
      unique_id:                             <computed>

  + aws_iam_role.qa_reports_role
      id:                                    <computed>
      arn:                                   <computed>
      assume_role_policy:                    "{\n    \"Version\": \"2012-10-17\",\n    \"Statement\": [\n        {\n            \"Effect\": \"Allow\",\n            \"Action\": \"sts:AssumeRole\",\n            \"Principal\": {\n               \"Service\": \"ec2.amazonaws.com\"\n            },\n            \"Sid\": \"\"\n        }\n    ]\n}\n"
      create_date:                           <computed>
      force_detach_policies:                 "false"
      max_session_duration:                  "3600"
      name:                                  "production_qa_reports_instance_iam_role"
      path:                                  "/"
      unique_id:                             <computed>

  + aws_iam_role_policy.qa_reports_role_policy
      id:                                    <computed>
      name:                                  "production_qa_reports_role_policy"
      policy:                                "{\n    \"Version\": \"2012-10-17\",\n    \"Statement\": [\n        {\n            \"Effect\": \"Allow\",\n            \"Action\": [\n                \"cloudwatch:PutMetricData\",\n                \"ec2:DescribeVolumes\",\n                \"ec2:DescribeTags\",\n                \"logs:PutLogEvents\",\n                \"logs:DescribeLogStreams\",\n                \"logs:DescribeLogGroups\",\n                \"logs:CreateLogStream\",\n                \"logs:CreateLogGroup\"\n            ],\n            \"Resource\": \"*\"\n        },\n        {\n            \"Effect\": \"Allow\",\n            \"Action\": [\n                \"ssm:GetParameter\"\n            ],\n            \"Resource\": \"arn:aws:ssm:*:*:parameter/AmazonCloudWatch-*\"\n        }\n    ]\n}\n"
      role:                                  "production_qa_reports_instance_iam_role"

  ~ module.rds.aws_db_parameter_group.default
      parameter.#:                           "1" => "9"
      parameter.124058312.apply_method:      "" => "immediate"
      parameter.124058312.name:              "" => "checkpoint_timeout"
      parameter.124058312.value:             "" => "1800"
      parameter.1764331501.apply_method:     "immediate" => "immediate"
      parameter.1764331501.name:             "log_min_duration_statement" => "log_min_duration_statement"
      parameter.1764331501.value:            "500" => "500"
      parameter.1790424064.apply_method:     "" => "immediate"
      parameter.1790424064.name:             "" => "log_statement"
      parameter.1790424064.value:            "" => "none"
      parameter.2060294046.apply_method:     "" => "immediate"
      parameter.2060294046.name:             "" => "max_wal_size"
      parameter.2060294046.value:            "" => "256"
      parameter.2771588648.apply_method:     "" => "immediate"
      parameter.2771588648.name:             "" => "synchronous_commit"
      parameter.2771588648.value:            "" => "off"
      parameter.3490834664.apply_method:     "" => "immediate"
      parameter.3490834664.name:             "" => "maintenance_work_mem"
      parameter.3490834664.value:            "" => "131072"
      parameter.490699910.apply_method:      "" => "immediate"
      parameter.490699910.name:              "" => "auto_explain.log_min_duration"
      parameter.490699910.value:             "" => "500"
      parameter.512379266.apply_method:      "" => "immediate"
      parameter.512379266.name:              "" => "wal_buffers"
      parameter.512379266.value:             "" => "8192"
      parameter.584376853.apply_method:      "" => "immediate"
      parameter.584376853.name:              "" => "autovacuum"
      parameter.584376853.value:             "" => "0"

  + module.sqs.aws_iam_role_policy.sqs_manage_policy
      id:                                    <computed>
      name:                                  "production_sqs_manage_policy"
      policy:                                "{\n    \"Version\": \"2012-10-17\",\n    \"Statement\": [\n        {\n            \"Sid\": \"VisualEditor0\",\n            \"Effect\": \"Allow\",\n            \"Action\": [\n                \"sqs:DeleteMessage\",\n                \"sqs:SendMessage\",\n                \"sqs:ReceiveMessage\",\n                \"sqs:GetQueueAttributes\",\n                \"sqs:GetQueueUrl\"\n            ],\n            \"Resource\": [\"${join(\"\\\",\\\"\", aws_sqs_queue.qa_reports_queue.*.arn)}\"]\n        },\n        {\n            \"Sid\": \"VisualEditor1\",\n            \"Effect\": \"Allow\",\n            \"Action\": \"sqs:ListQueues\",\n            \"Resource\": \"*\"\n        }\n    ]\n}\n"
      role:                                  "production_qa_reports_instance_iam_role"

  + module.sqs.aws_sqs_queue.qa_reports_queue[0]
      id:                                    <computed>
      arn:                                   <computed>
      content_based_deduplication:           "false"
      delay_seconds:                         "0"
      fifo_queue:                            "false"
      kms_data_key_reuse_period_seconds:     <computed>
      max_message_size:                      "262144"
      message_retention_seconds:             "345600"
      name:                                  "production_celery"
      policy:                                <computed>
      receive_wait_time_seconds:             "0"
      visibility_timeout_seconds:            "30"

  + module.sqs.aws_sqs_queue.qa_reports_queue[1]
      id:                                    <computed>
      arn:                                   <computed>
      content_based_deduplication:           "false"
      delay_seconds:                         "0"
      fifo_queue:                            "false"
      kms_data_key_reuse_period_seconds:     <computed>
      max_message_size:                      "262144"
      message_retention_seconds:             "345600"
      name:                                  "production_ci_fetch"
      policy:                                <computed>
      receive_wait_time_seconds:             "0"
      visibility_timeout_seconds:            "30"

  + module.sqs.aws_sqs_queue.qa_reports_queue[2]
      id:                                    <computed>
      arn:                                   <computed>
      content_based_deduplication:           "false"
      delay_seconds:                         "0"
      fifo_queue:                            "false"
      kms_data_key_reuse_period_seconds:     <computed>
      max_message_size:                      "262144"
      message_retention_seconds:             "345600"
      name:                                  "production_ci_poll"
      policy:                                <computed>
      receive_wait_time_seconds:             "0"
      visibility_timeout_seconds:            "30"

  + module.sqs.aws_sqs_queue.qa_reports_queue[3]
      id:                                    <computed>
      arn:                                   <computed>
      content_based_deduplication:           "false"
      delay_seconds:                         "0"
      fifo_queue:                            "false"
      kms_data_key_reuse_period_seconds:     <computed>
      max_message_size:                      "262144"
      message_retention_seconds:             "345600"
      name:                                  "production_ci_quick"
      policy:                                <computed>
      receive_wait_time_seconds:             "0"
      visibility_timeout_seconds:            "30"

  + module.sqs.aws_sqs_queue.qa_reports_queue[4]
      id:                                    <computed>
      arn:                                   <computed>
      content_based_deduplication:           "false"
      delay_seconds:                         "0"
      fifo_queue:                            "false"
      kms_data_key_reuse_period_seconds:     <computed>
      max_message_size:                      "262144"
      message_retention_seconds:             "345600"
      name:                                  "production_core_notification"
      policy:                                <computed>
      receive_wait_time_seconds:             "0"
      visibility_timeout_seconds:            "30"

  + module.sqs.aws_sqs_queue.qa_reports_queue[5]
      id:                                    <computed>
      arn:                                   <computed>
      content_based_deduplication:           "false"
      delay_seconds:                         "0"
      fifo_queue:                            "false"
      kms_data_key_reuse_period_seconds:     <computed>
      max_message_size:                      "262144"
      message_retention_seconds:             "345600"
      name:                                  "production_core_postprocess"
      policy:                                <computed>
      receive_wait_time_seconds:             "0"
      visibility_timeout_seconds:            "30"

  + module.sqs.aws_sqs_queue.qa_reports_queue[6]
      id:                                    <computed>
      arn:                                   <computed>
      content_based_deduplication:           "false"
      delay_seconds:                         "0"
      fifo_queue:                            "false"
      kms_data_key_reuse_period_seconds:     <computed>
      max_message_size:                      "262144"
      message_retention_seconds:             "345600"
      name:                                  "production_core_quick"
      policy:                                <computed>
      receive_wait_time_seconds:             "0"
      visibility_timeout_seconds:            "30"

  + module.sqs.aws_sqs_queue.qa_reports_queue[7]
      id:                                    <computed>
      arn:                                   <computed>
      content_based_deduplication:           "false"
      delay_seconds:                         "0"
      fifo_queue:                            "false"
      kms_data_key_reuse_period_seconds:     <computed>
      max_message_size:                      "262144"
      message_retention_seconds:             "345600"
      name:                                  "production_core_reporting"
      policy:                                <computed>
      receive_wait_time_seconds:             "0"
      visibility_timeout_seconds:            "30"

  ~ module.webservers.aws_cloudwatch_metric_alarm.worker_disk_usage[0]
      dimensions.%:                          "" => <computed>
      dimensions.InstanceId:                 "i-0ebd191d546eac7af" => ""
      dimensions.device:                     "xvda1" => ""
      dimensions.fstype:                     "ext4" => ""
      dimensions.path:                       "/" => ""

  ~ module.webservers.aws_cloudwatch_metric_alarm.worker_disk_usage[1]
      dimensions.%:                          "" => <computed>
      dimensions.InstanceId:                 "i-07df711da0c6e9320" => ""
      dimensions.device:                     "xvda1" => ""
      dimensions.fstype:                     "ext4" => ""
      dimensions.path:                       "/" => ""

  ~ module.webservers.aws_cloudwatch_metric_alarm.worker_disk_usage[2]
      dimensions.%:                          "" => <computed>
      dimensions.InstanceId:                 "i-07ea0b70b54c08ace" => ""
      dimensions.device:                     "xvda1" => ""
      dimensions.fstype:                     "ext4" => ""
      dimensions.path:                       "/" => ""

  ~ module.webservers.aws_cloudwatch_metric_alarm.www_disk_usage[0]
      dimensions.%:                          "" => <computed>
      dimensions.InstanceId:                 "i-07afe5562c94db883" => ""
      dimensions.device:                     "xvda1" => ""
      dimensions.fstype:                     "ext4" => ""
      dimensions.path:                       "/" => ""

  ~ module.webservers.aws_cloudwatch_metric_alarm.www_disk_usage[1]
      dimensions.%:                          "" => <computed>
      dimensions.InstanceId:                 "i-0c703fe5a346fb02f" => ""
      dimensions.device:                     "xvda1" => ""
      dimensions.fstype:                     "ext4" => ""
      dimensions.path:                       "/" => ""

-/+ module.webservers.aws_instance.qa-reports-worker[0] (new resource required)
      id:                                    "i-0ebd191d546eac7af" => <computed> (forces new resource)
      ami:                                   "ami-0b383171" => "ami-0b383171"
      arn:                                   "arn:aws:ec2:us-east-1:495570029050:instance/i-0ebd191d546eac7af" => <computed>
      associate_public_ip_address:           "true" => <computed>
      availability_zone:                     "us-east-1a" => "us-east-1a"
      cpu_core_count:                        "2" => <computed>
      cpu_threads_per_core:                  "1" => <computed>
      ebs_block_device.#:                    "0" => <computed>
      ephemeral_block_device.#:              "0" => <computed>
      get_password_data:                     "false" => "false"
      host_id:                               "" => <computed>
      iam_instance_profile:                  "qa_reports_instance_profile" => "production_qa_reports_instance_profile"
      instance_state:                        "running" => <computed>
      instance_type:                         "t2.medium" => "t2.medium"
      ipv6_address_count:                    "0" => <computed>
      ipv6_addresses.#:                      "0" => <computed>
      key_name:                              "qa-reports-production" => "qa-reports-production"
      network_interface.#:                   "0" => <computed>
      network_interface_id:                  "eni-0026186b6b795aff9" => <computed>
      password_data:                         "" => <computed>
      placement_group:                       "" => <computed>
      primary_network_interface_id:          "eni-0026186b6b795aff9" => <computed>
      private_dns:                           "ip-172-31-1-139.ec2.internal" => <computed>
      private_ip:                            "172.31.1.139" => <computed>
      public_dns:                            "ec2-18-212-9-17.compute-1.amazonaws.com" => <computed>
      public_ip:                             "18.212.9.17" => <computed>
      root_block_device.#:                   "1" => <computed>
      security_groups.#:                     "0" => <computed>
      source_dest_check:                     "true" => "true"
      subnet_id:                             "subnet-08375ce285b190f06" => "subnet-08375ce285b190f06"
      tags.%:                                "1" => "1"
      tags.Name:                             "qa-reports-worker-0" => "qa-reports-worker-0"
      tenancy:                               "default" => <computed>
      user_data:                             "34fb99e18d8530d5e6d2487b968bfdcbad274823" => "42ae120c25fda6907d7ac34ea6b22a8f236b0340" (forces new resource)
      volume_tags.%:                         "1" => <computed>
      vpc_security_group_ids.#:              "1" => "1"
      vpc_security_group_ids.909090015:      "sg-0b4879d45d89a3d0b" => "sg-0b4879d45d89a3d0b"

-/+ module.webservers.aws_instance.qa-reports-worker[1] (new resource required)
      id:                                    "i-07df711da0c6e9320" => <computed> (forces new resource)
      ami:                                   "ami-0b383171" => "ami-0b383171"
      arn:                                   "arn:aws:ec2:us-east-1:495570029050:instance/i-07df711da0c6e9320" => <computed>
      associate_public_ip_address:           "true" => <computed>
      availability_zone:                     "us-east-1b" => "us-east-1b"
      cpu_core_count:                        "2" => <computed>
      cpu_threads_per_core:                  "1" => <computed>
      ebs_block_device.#:                    "0" => <computed>
      ephemeral_block_device.#:              "0" => <computed>
      get_password_data:                     "false" => "false"
      host_id:                               "" => <computed>
      iam_instance_profile:                  "qa_reports_instance_profile" => "production_qa_reports_instance_profile"
      instance_state:                        "running" => <computed>
      instance_type:                         "t2.medium" => "t2.medium"
      ipv6_address_count:                    "0" => <computed>
      ipv6_addresses.#:                      "0" => <computed>
      key_name:                              "qa-reports-production" => "qa-reports-production"
      network_interface.#:                   "0" => <computed>
      network_interface_id:                  "eni-0e6f69a2b6243a757" => <computed>
      password_data:                         "" => <computed>
      placement_group:                       "" => <computed>
      primary_network_interface_id:          "eni-0e6f69a2b6243a757" => <computed>
      private_dns:                           "ip-172-31-2-150.ec2.internal" => <computed>
      private_ip:                            "172.31.2.150" => <computed>
      public_dns:                            "ec2-3-82-25-12.compute-1.amazonaws.com" => <computed>
      public_ip:                             "3.82.25.12" => <computed>
      root_block_device.#:                   "1" => <computed>
      security_groups.#:                     "0" => <computed>
      source_dest_check:                     "true" => "true"
      subnet_id:                             "subnet-0197bdcd7f985caa4" => "subnet-0197bdcd7f985caa4"
      tags.%:                                "1" => "1"
      tags.Name:                             "qa-reports-worker-1" => "qa-reports-worker-1"
      tenancy:                               "default" => <computed>
      user_data:                             "34fb99e18d8530d5e6d2487b968bfdcbad274823" => "42ae120c25fda6907d7ac34ea6b22a8f236b0340" (forces new resource)
      volume_tags.%:                         "1" => <computed>
      vpc_security_group_ids.#:              "1" => "1"
      vpc_security_group_ids.909090015:      "sg-0b4879d45d89a3d0b" => "sg-0b4879d45d89a3d0b"

  + module.webservers.aws_instance.qa-reports-worker[2]
      id:                                    <computed>
      ami:                                   "ami-0b383171"
      arn:                                   <computed>
      associate_public_ip_address:           <computed>
      availability_zone:                     "us-east-1a"
      cpu_core_count:                        <computed>
      cpu_threads_per_core:                  <computed>
      ebs_block_device.#:                    <computed>
      ephemeral_block_device.#:              <computed>
      get_password_data:                     "false"
      host_id:                               <computed>
      iam_instance_profile:                  "production_qa_reports_instance_profile"
      instance_state:                        <computed>
      instance_type:                         "t2.medium"
      ipv6_address_count:                    <computed>
      ipv6_addresses.#:                      <computed>
      key_name:                              "qa-reports-production"
      network_interface.#:                   <computed>
      network_interface_id:                  <computed>
      password_data:                         <computed>
      placement_group:                       <computed>
      primary_network_interface_id:          <computed>
      private_dns:                           <computed>
      private_ip:                            <computed>
      public_dns:                            <computed>
      public_ip:                             <computed>
      root_block_device.#:                   <computed>
      security_groups.#:                     <computed>
      source_dest_check:                     "true"
      subnet_id:                             "subnet-08375ce285b190f06"
      tags.%:                                "1"
      tags.Name:                             "qa-reports-worker-2"
      tenancy:                               <computed>
      user_data:                             "42ae120c25fda6907d7ac34ea6b22a8f236b0340"
      volume_tags.%:                         <computed>
      vpc_security_group_ids.#:              "1"
      vpc_security_group_ids.909090015:      "sg-0b4879d45d89a3d0b"

-/+ module.webservers.aws_instance.qa-reports-www[0] (new resource required)
      id:                                    "i-07afe5562c94db883" => <computed> (forces new resource)
      ami:                                   "ami-0b383171" => "ami-0b383171"
      arn:                                   "arn:aws:ec2:us-east-1:495570029050:instance/i-07afe5562c94db883" => <computed>
      associate_public_ip_address:           "true" => <computed>
      availability_zone:                     "us-east-1a" => "us-east-1a"
      cpu_core_count:                        "2" => <computed>
      cpu_threads_per_core:                  "1" => <computed>
      ebs_block_device.#:                    "0" => <computed>
      ephemeral_block_device.#:              "0" => <computed>
      get_password_data:                     "false" => "false"
      host_id:                               "" => <computed>
      iam_instance_profile:                  "qa_reports_instance_profile" => "production_qa_reports_instance_profile"
      instance_state:                        "running" => <computed>
      instance_type:                         "t2.medium" => "t2.medium"
      ipv6_address_count:                    "0" => <computed>
      ipv6_addresses.#:                      "0" => <computed>
      key_name:                              "qa-reports-production" => "qa-reports-production"
      network_interface.#:                   "0" => <computed>
      network_interface_id:                  "eni-08059c2b4f1e10061" => <computed>
      password_data:                         "" => <computed>
      placement_group:                       "" => <computed>
      primary_network_interface_id:          "eni-08059c2b4f1e10061" => <computed>
      private_dns:                           "ip-172-31-1-92.ec2.internal" => <computed>
      private_ip:                            "172.31.1.92" => <computed>
      public_dns:                            "ec2-35-173-242-254.compute-1.amazonaws.com" => <computed>
      public_ip:                             "35.173.242.254" => <computed>
      root_block_device.#:                   "1" => <computed>
      security_groups.#:                     "0" => <computed>
      source_dest_check:                     "true" => "true"
      subnet_id:                             "subnet-08375ce285b190f06" => "subnet-08375ce285b190f06"
      tags.%:                                "1" => "1"
      tags.Name:                             "qa-reports-www-0" => "qa-reports-www-0"
      tenancy:                               "default" => <computed>
      user_data:                             "34fb99e18d8530d5e6d2487b968bfdcbad274823" => "42ae120c25fda6907d7ac34ea6b22a8f236b0340" (forces new resource)
      volume_tags.%:                         "1" => <computed>
      vpc_security_group_ids.#:              "1" => "1"
      vpc_security_group_ids.3267943901:     "sg-0784d01ac183ae01c" => "sg-0784d01ac183ae01c"

-/+ module.webservers.aws_instance.qa-reports-www[1] (new resource required)
      id:                                    "i-0c703fe5a346fb02f" => <computed> (forces new resource)
      ami:                                   "ami-0b383171" => "ami-0b383171"
      arn:                                   "arn:aws:ec2:us-east-1:495570029050:instance/i-0c703fe5a346fb02f" => <computed>
      associate_public_ip_address:           "true" => <computed>
      availability_zone:                     "us-east-1b" => "us-east-1b"
      cpu_core_count:                        "2" => <computed>
      cpu_threads_per_core:                  "1" => <computed>
      ebs_block_device.#:                    "0" => <computed>
      ephemeral_block_device.#:              "0" => <computed>
      get_password_data:                     "false" => "false"
      host_id:                               "" => <computed>
      iam_instance_profile:                  "qa_reports_instance_profile" => "production_qa_reports_instance_profile"
      instance_state:                        "running" => <computed>
      instance_type:                         "t2.medium" => "t2.medium"
      ipv6_address_count:                    "0" => <computed>
      ipv6_addresses.#:                      "0" => <computed>
      key_name:                              "qa-reports-production" => "qa-reports-production"
      network_interface.#:                   "0" => <computed>
      network_interface_id:                  "eni-0c0dd1bd6ff300b37" => <computed>
      password_data:                         "" => <computed>
      placement_group:                       "" => <computed>
      primary_network_interface_id:          "eni-0c0dd1bd6ff300b37" => <computed>
      private_dns:                           "ip-172-31-2-241.ec2.internal" => <computed>
      private_ip:                            "172.31.2.241" => <computed>
      public_dns:                            "ec2-34-224-70-97.compute-1.amazonaws.com" => <computed>
      public_ip:                             "34.224.70.97" => <computed>
      root_block_device.#:                   "1" => <computed>
      security_groups.#:                     "0" => <computed>
      source_dest_check:                     "true" => "true"
      subnet_id:                             "subnet-0197bdcd7f985caa4" => "subnet-0197bdcd7f985caa4"
      tags.%:                                "1" => "1"
      tags.Name:                             "qa-reports-www-1" => "qa-reports-www-1"
      tenancy:                               "default" => <computed>
      user_data:                             "34fb99e18d8530d5e6d2487b968bfdcbad274823" => "42ae120c25fda6907d7ac34ea6b22a8f236b0340" (forces new resource)
      volume_tags.%:                         "1" => <computed>
      vpc_security_group_ids.#:              "1" => "1"
      vpc_security_group_ids.3267943901:     "sg-0784d01ac183ae01c" => "sg-0784d01ac183ae01c"

-/+ module.webservers.aws_lb_target_group_attachment.qa-reports-www-lb[0] (new resource required)
      id:                                    "arn:aws:elasticloadbalancing:us-east-1:495570029050:targetgroup/qa-reports-tg/0825c86d0387f9fd-20180420150016760500000002" => <computed> (forces new resource)
      port:                                  "80" => "80"
      target_group_arn:                      "arn:aws:elasticloadbalancing:us-east-1:495570029050:targetgroup/qa-reports-tg/0825c86d0387f9fd" => "arn:aws:elasticloadbalancing:us-east-1:495570029050:targetgroup/qa-reports-tg/0825c86d0387f9fd"
      target_id:                             "i-07afe5562c94db883" => "${element(aws_instance.qa-reports-www.*.id, count.index)}" (forces new resource)

-/+ module.webservers.aws_lb_target_group_attachment.qa-reports-www-lb[1] (new resource required)
      id:                                    "arn:aws:elasticloadbalancing:us-east-1:495570029050:targetgroup/qa-reports-tg/0825c86d0387f9fd-20180420150016829100000003" => <computed> (forces new resource)
      port:                                  "80" => "80"
      target_group_arn:                      "arn:aws:elasticloadbalancing:us-east-1:495570029050:targetgroup/qa-reports-tg/0825c86d0387f9fd" => "arn:aws:elasticloadbalancing:us-east-1:495570029050:targetgroup/qa-reports-tg/0825c86d0387f9fd"
      target_id:                             "i-0c703fe5a346fb02f" => "${element(aws_instance.qa-reports-www.*.id, count.index)}" (forces new resource)

  ~ module.webservers.aws_security_group.qa-reports-ec2-www
      ingress.#:                             "7" => "4"
      ingress.1732791625.cidr_blocks.#:      "2" => "0"
      ingress.1732791625.cidr_blocks.0:      "172.31.1.0/24" => ""
      ingress.1732791625.cidr_blocks.1:      "172.31.2.0/24" => ""
      ingress.1732791625.description:        "" => ""
      ingress.1732791625.from_port:          "5671" => "0"
      ingress.1732791625.ipv6_cidr_blocks.#: "0" => "0"
      ingress.1732791625.prefix_list_ids.#:  "0" => "0"
      ingress.1732791625.protocol:           "tcp" => ""
      ingress.1732791625.security_groups.#:  "0" => "0"
      ingress.1732791625.self:               "false" => "false"
      ingress.1732791625.to_port:            "5672" => "0"
      ingress.2541437006.cidr_blocks.#:      "1" => "1"
      ingress.2541437006.cidr_blocks.0:      "0.0.0.0/0" => "0.0.0.0/0"
      ingress.2541437006.description:        "" => ""
      ingress.2541437006.from_port:          "22" => "22"
      ingress.2541437006.ipv6_cidr_blocks.#: "0" => "0"
      ingress.2541437006.prefix_list_ids.#:  "0" => "0"
      ingress.2541437006.protocol:           "tcp" => "tcp"
      ingress.2541437006.security_groups.#:  "0" => "0"
      ingress.2541437006.self:               "false" => "false"
      ingress.2541437006.to_port:            "22" => "22"
      ingress.2579333545.cidr_blocks.#:      "2" => "0"
      ingress.2579333545.cidr_blocks.0:      "172.31.1.0/24" => ""
      ingress.2579333545.cidr_blocks.1:      "172.31.2.0/24" => ""
      ingress.2579333545.description:        "" => ""
      ingress.2579333545.from_port:          "4369" => "0"
      ingress.2579333545.ipv6_cidr_blocks.#: "0" => "0"
      ingress.2579333545.prefix_list_ids.#:  "0" => "0"
      ingress.2579333545.protocol:           "tcp" => ""
      ingress.2579333545.security_groups.#:  "0" => "0"
      ingress.2579333545.self:               "false" => "false"
      ingress.2579333545.to_port:            "4369" => "0"
      ingress.2803363525.cidr_blocks.#:      "2" => "0"
      ingress.2803363525.cidr_blocks.0:      "172.31.1.0/24" => ""
      ingress.2803363525.cidr_blocks.1:      "172.31.2.0/24" => ""
      ingress.2803363525.description:        "" => ""
      ingress.2803363525.from_port:          "25672" => "0"
      ingress.2803363525.ipv6_cidr_blocks.#: "0" => "0"
      ingress.2803363525.prefix_list_ids.#:  "0" => "0"
      ingress.2803363525.protocol:           "tcp" => ""
      ingress.2803363525.security_groups.#:  "0" => "0"
      ingress.2803363525.self:               "false" => "false"
      ingress.2803363525.to_port:            "25672" => "0"
      ingress.3526486982.cidr_blocks.#:      "2" => "2"
      ingress.3526486982.cidr_blocks.0:      "172.31.1.0/24" => "172.31.1.0/24"
      ingress.3526486982.cidr_blocks.1:      "172.31.2.0/24" => "172.31.2.0/24"
      ingress.3526486982.description:        "" => ""
      ingress.3526486982.from_port:          "4949" => "4949"
      ingress.3526486982.ipv6_cidr_blocks.#: "0" => "0"
      ingress.3526486982.prefix_list_ids.#:  "0" => "0"
      ingress.3526486982.protocol:           "tcp" => "tcp"
      ingress.3526486982.security_groups.#:  "0" => "0"
      ingress.3526486982.self:               "false" => "false"
      ingress.3526486982.to_port:            "4949" => "4949"
      ingress.4222347986.cidr_blocks.#:      "2" => "2"
      ingress.4222347986.cidr_blocks.0:      "172.31.1.0/24" => "172.31.1.0/24"
      ingress.4222347986.cidr_blocks.1:      "172.31.2.0/24" => "172.31.2.0/24"
      ingress.4222347986.description:        "" => ""
      ingress.4222347986.from_port:          "19532" => "19532"
      ingress.4222347986.ipv6_cidr_blocks.#: "0" => "0"
      ingress.4222347986.prefix_list_ids.#:  "0" => "0"
      ingress.4222347986.protocol:           "tcp" => "tcp"
      ingress.4222347986.security_groups.#:  "0" => "0"
      ingress.4222347986.self:               "false" => "false"
      ingress.4222347986.to_port:            "19532" => "19532"
      ingress.865157667.cidr_blocks.#:       "2" => "2"
      ingress.865157667.cidr_blocks.0:       "172.31.1.0/24" => "172.31.1.0/24"
      ingress.865157667.cidr_blocks.1:       "172.31.2.0/24" => "172.31.2.0/24"
      ingress.865157667.description:         "" => ""
      ingress.865157667.from_port:           "80" => "80"
      ingress.865157667.ipv6_cidr_blocks.#:  "0" => "0"
      ingress.865157667.prefix_list_ids.#:   "0" => "0"
      ingress.865157667.protocol:            "tcp" => "tcp"
      ingress.865157667.security_groups.#:   "0" => "0"
      ingress.865157667.self:                "false" => "false"
      ingress.865157667.to_port:             "80" => "80"


Plan: 19 to add, 7 to change, 6 to destroy.
```

danrue commented:
Recreating all the instances like that will be risky and cause some downtime. I'm not sure of an easy way to reduce the risk - I was hoping for some way to defer the delete so that you could validate the new systems before destroying the old ones, but I don't see any unless you write blue/green logic into your terraform itself. Alternatively, doing the apply step-wise would also work - where it pauses before each step for validation. There is a -parallelism option that could be set to 1, but i don't know how to have it wait. Thoughts?

chaws commented:
@danrue sorry for not replying to this before.

I've checked a few posts on how to implement blue/green with terraform and it seemed too complex to what we want to achieve right now. Because we're also planning to migrate our infra to ecs in the next couple of months, we opted for simplicity: 

* I'll detach all 5 ec2 instances from terraform state file
* then I'll run `apply` at a time that the queue is mostly quiet (around midnight UTC-3h). This way, while the new infrastructure is being created, the old one will still finish whatever was in the old rabbitmq queue. It should be a downtime of a few minutes. I'll send out and email with the announcement

chaws committed:
terraform: scripts: provision: add Charles ssh key

chaws committed:
ansible: update hosts ip addresses
